### PR TITLE
[CI/CD] Introduce the `test_publish_github_release` workflow

### PR DESCRIPTION
**Pull request by alexander-smolyakov**
**Description**:
This PR introduces the `test_publish_github_release` workflow to test GitHub Release workflows in this repo.

_Changelog_:
- Introduces the `test_publish_github_release` workflow;
- The `build-test-package.yml` workflow converted to reusable workflow;
- Bumped up version for several GH actions to resolve warnings and reduce noise in the annotation section on the workflow run page.

### CODE REVIEW

File: .github/workflows/build-test-package.yml [hunk 1]
```diff
@@ -107,15 +109,15 @@ jobs:
           retention-days: ${{ env.ARTIFACT_RETENTION_DAYS }}
 
   test-build:
-    name: verify packages
+    name: Verify packages
 
-    needs: [build, unit]
```
Comment (emmyoop) [inline .github/workflows/build-test-package.yml:112]: Out of curiosity why did you change the order of there?

File: .github/workflows/test_publish_github_release.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+  build-test-verify-publish-package:
+    if: inputs.generate_build_artifact
+    uses: dbt-labs/dbt-core-release-test/.github/workflows/build-test-package.yml@main
+    with:
+      sha: ${{ inputs.sha }}
+      version_number: ${{ inputs.version_number }}
+
+  publish-github-release:
+    if: ${{ always() && !cancelled() }}
```
Comment (emmyoop) [inline .github/workflows/test_publish_github_release.yml:48]: Why would we only want to skip this job when `build-test-verify-publish-package` was cancelled?  Shouldn't we also stop on failure of that job?

**Review by emmyoop — COMMENTED** [review]
A few questions, but mostly looks good.

File: .github/workflows/test_publish_github_release.yml (hunk 2 above)
Comment (alexander-smolyakov) [reply .github/workflows/test_publish_github_release.yml:48]: This workflow is for testing purposes only. I want to be able to test the behavior of workflow even in cases where pre-GitHub release workflow steps failed.

File: .github/workflows/build-test-package.yml (hunk 1 above)
Comment (alexander-smolyakov) [reply .github/workflows/build-test-package.yml:112]: Just wanted to align them with the order their defined in the YAML definition.

**Review by emmyoop — APPROVED** [review]
